socialhome/streams/app/components/AuthorBar.vue

```diff
@@ -53,7 +53,7 @@ import Vue from "vue"
 
 export default Vue.component("author-bar", {
     props: {
-        contentId: {type: Number, required: true},
+        content: {type: Object, required: true},
```

Comment:
Why not use the store directly?

Reply:
Because whatever is passed to `StreamElement` and the below components down the line could be either in `state.contents` or `state.replies`. Since these were separated to make the store cleaner, it means the components must receive the object directly to avoid if/else in the properties (and passing an extra type to the component).

Reply:
Can you then specify each property so that we benefit from type checking? If you need to modify properties, you can use the [`.sync` modifier](https://vuejs.org/v2/guide/components.html#sync-Modifier) on `v-bind`.

Reply:
I'll have a look at that, good point. Though only required ones should be given, otherwise we break the UI every time the `Content` serializer changes - which is every week currently ;)

Reply:
Seems to me it's not possible to define a validator for `object` requiring only certain keys to be present. One possibility is to write a [validator function](https://vuejs.org/v2/guide/components.html#Prop-Validation). I'm not honestly sure it's worth the effort since the function would be unique for each component. Some of the components require different properties available on the object.

Reply:
If the object will be that different whether it's a post or a reply, then maybe it is worth writing two seperate templates. Otherwise, just detail all the properties and bind the whole object using `v-bind` ;)

Reply:
The whole point of reusing the `StreamElement` component is to ensure we don't *have to* write separate components that will be 98% the same lines of code.

Can you give an example of `v-bind` in this case? AFAICT we are already doing it by passing in the content with the `v-bind` shorthand `:content` from the parent. But I might be [misunderstanding the docs](https://vuejs.org/v2/guide/components.html#Dynamic-Props) :)

Reply:
```js
props: {
  author: {type: Object, required: true},
  user_is_author: {type: Boolean, required: true}
  user_is_following_author: {type: Boolean, required: true}
  // ...
}
```
and
```js
<author-bar v-bind="content"/>
```
should do.

Reply:
Also, at some point, I'd like we speak again about snake case in the API. Make the whole front-end inconsistent and it's not common. Most HTTP APIs I've seen so far always used camel case. There must be some option to make Django REST API deliver camel case.

Reply:
Interesting, have to try this tonight at home. Basically, we're already using `v-bind` as follows (`:content="content"` is the same as `v-bind="content"` according to the docs.. I mean, why have one way of writing something when you can have many!).

So how does the current code even work? Looking at your example [this should crash](https://github.com/jaywink/socialhome/blob/vue-replies/socialhome/streams/app/components/AuthorBar.vue#L56), but it works fine. I'll see what happens if I extract the object properties.

I'm concerned however that doing it that way will create more mess and less clarity. For example, if we reference `id` in `AuthorBar`, how can the reader know it's Content ID. I would myself assume Author ID just by looking at the component. Passing in `contentId` as before or `content` as in the new code makes it clear what is being referenced. Having `id` as a property is vague.

Regarding the camelCase in the REST API. I'm afraid this is a no. The reasons are the same:

* There is [no standard](https://stackoverflow.com/a/25368854/1489738) in [JSON](https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf) that objects should have snake_case or camelCase attributes.
* Similarly there is no standard for REST either (well, it's not really a data format specification...)
* The Socialhome API is not just for the frontend. And not just for JavaScript consumers.
* Adding camelCase to REST framework is only possible via a third-party dependency, as you said. It is not justified to add new dependencies just because the JS code "looks bad". The extra dependency could some day stop us from upgrading to a newer DRF version, for example, or could [contain bugs](https://github.com/vbabiy/djangorestframework-camel-case/issues?utf8=%E2%9C%93&q=is%3Aissue).

Don't worry, the snakes wont bite ;)

Reply:
> Basically, we're already using `v-bind` as follows (`:content="content"` is the same as `v-bind="content"` according to the docs.

No, `:content="content"` is a shorthand for `v-bind:content="content"`. `v-bind="content"` binds `content` properties to matching component properties.

> Don't worry, the snakes wont bite ;)
 
No, but in this case, it forces front-end to write some properties in snake case and other in camel case :/

Reply:
Yeah the `v-bind="content"` works 👍 Thanks!

Now, the only problem is, I'm really not sure this is a good way to go. What is the actual benefit we're trying to achieve here except making code more verbose? We're using pretty much objects coming from the API. If the API changes, tests should fail anyway. We can't rely on devs reading console outputs, which is the only thing props validation really adds. I really can't see why we should add more verbosity to the calling code and the tests just because we can.

I'm open to consider doing this where it makes sense, for example in `AuthorBar` the properties used from `content` object kind of make sense as individual attributes. But in some of the receiving components they stop making sense individually once extracted from the `content`. For example, if we want to pass in an attribute called `id`, then the component name should say what the ID is. I don't particularly like the idea of generating extra verbose code and having longer component names just to get dev console warnings.

I'm not saying we should not add props validation when we have clear attributes that we pass in, but if it makes *sense* to pass in an object, we shouldn't add extra verbosity by ripping the object apart creating confusing out of context attributes on the receiving component.

Reply:
To get forward with this, I'm going to merge once tests pass. We can think of using the `v-bind` case by case according to above notes. In these cases I don't think it makes sense because:

* it increases verbosity with little benefit (only a console warning, having tests fail would actually be a little more useful)
* it can provide confusing variable names to local data
* it doesn't even allow validation objects with sub-objects, for example when passing `v-bind="content"` our `props` would still have to have `author: object`

Still, thanks for raising this! Let's keep the option case by case 👍